pymongo/auth_oidc.py

```diff
@@ -247,37 +173,59 @@ def auth_start_cmd(self, use_callbacks: bool = True) -> Optional[SON[str, Any]]:
             ]
         )
 
-    def clear(self) -> None:
-        self.idp_info = None
-        self.idp_resp = None
-        self.token_exp_utc = None
-
     def run_command(
         self, conn: Connection, cmd: MutableMapping[str, Any]
     ) -> Optional[Mapping[str, Any]]:
         try:
             return conn.command("$external", cmd, no_reauth=True)  # type: ignore[call-arg]
-        except OperationFailure as exc:
-            self.clear()
-            if exc.code == _REAUTHENTICATION_REQUIRED_CODE:
-                if "jwt" in bson.decode(cmd["payload"]):
-                    if self.idp_info_gen_id > self.reauth_gen_id:
-                        raise
-                    return self.authenticate(conn, reauthenticate=True)
+        except OperationFailure:
+            self.access_token = None
             raise
 
-    def authenticate(
-        self, conn: Connection, reauthenticate: bool = False
-    ) -> Optional[Mapping[str, Any]]:
-        if reauthenticate:
-            prev_id = getattr(conn, "oidc_token_gen_id", None)
-            # Check if we've already changed tokens.
-            if prev_id == self.token_gen_id:
-                self.reauth_gen_id = self.idp_info_gen_id
-                self.token_exp_utc = None
-                if not self.properties.refresh_token_callback:
-                    self.clear()
+    def reauthenticate(self, conn: Connection) -> Optional[Mapping[str, Any]]:
+        """Handle a reauthenticate from the server."""
+        # First see if we have the a newer token on the authenticator.
+        prev_id = getattr(conn, "oidc_token_gen_id", 0)
```

Comment:
Let's make oidc_token_gen_id a well defined property on Connection and avoid getattr.

Reply:
Ah, it was already there as an optional int, updated